### PR TITLE
Capstone project: Tunisia Covid Tracker

### PR DESCRIPTION
**Pull request by taher-software**
In this project , I have created a mobile web application to track Covid situation in my country Tunisia. This application display the main indicators and metrics allow to grasp the real situation in a country, it ensure also the running of all the features required in this capstone project. Therefore you find in this branch the following changes : :100: 
- Home page : 
     * When the page loads, the webapp shows the list of main indicators  that could give a general idea about the country Covid spread situation. :+1: 
    * Along with the category name, we display some numeric values per category that come from the API. :+1: 
    * When the user clicks (or taps) on a category item, the application navigates to the details page. :+1: 
- Details page:
    * In the details page, the webapp retrieves data from the API to show detailed data in the category. :+1: 
    * When the user clicks on the "Back" button (<), the user navigates to the home page. :+1: 
 
You can watch this video presentation of my project :
https://www.loom.com/share/cf1b20cd49504fd490a3d4157f398d42

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import ConfirmedCases from './components/ConfirmedCases/ConfirmedCases';
+import DeathsCases from './components/Deaths/Deaths';
+import RecoveredCases from './components/Recovered/Recovered';
+import OpenCases from './components/OpenCases/OpenCases';
+import './App.css';
+import lastDayNumbers from './redux/LATEST/thunk/thunk';
+import dayBeforeNumbers from './redux/DAYBEFORE/thunk/thunk';
+import getHistoricalData from './redux/HISTORICalMONTHDATA/thunk/thunk';
+
+function App() {
```
Comment (rukundoeric) [inline src/App.js]: - [ ] <b>REQUIRED:<b/> Prefer using ES6 arrow function. Actually, there is nothing wrong with ES5 functions syntax but it is mandatory to use es6 features in the javascript module. besides, Arrow functions provide a simplified structure, allowing for a more concise way to write function expressions. which makes them better than ES5 functions(function declarations). Here is the syntax of an ES6 arrow function...
```javascript
  const functionName = () => {
     // code here
  }
```

**Review by rukundoeric — CHANGES_REQUESTED** [review]
Hi @Taher-web-dev, 

_Great job so far 👏 💯 👍. However, there are few issues we need to address before we move on. Please, check the comments below._
## Highlights
  - Follows git workflow.  💯
  - Codebase looks clean and professional. 💯 

## Required Changes ♻️
  - [ ] You only have two test cases and you did not test other components except home. Kindly test other components and make sure to not only test `snapshots`.
  - [ ] Check the comments under the review.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
  - [ ] Check the comments under the review.

Cheers and Happy coding!👏👏👏

__Feel free to leave any questions or comments in the PR thread if something is not 100% clear.__

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
> As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by rukundoeric — CHANGES_REQUESTED** [review]
Hi @Taher-web-dev, 

_Great job so far 👏 💯 👍. You've done well implementing some of the requested changes. However, there are still some which aren't addressed yet. Please, check the comments below._
## Highlights
  - Follows git workflow.  💯
  - Codebase looks clean and professional. 💯 
  - Addressed some of the issues raised from the previous review. 💯 

## Required Changes ♻️
  - [ ] As I mentioned in the previous review, You only have two test cases and you did not test other components except home. Kindly test other components and make sure to not only test `snapshots`.
  Screen Shot 2021-12-23 at 15 46 06


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
  
  - N/A 

Cheers and Happy coding!👏👏👏

__Feel free to leave any questions or comments in the PR thread if something is not 100% clear.__

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
> As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: src/__test__/ChartOPenCases.test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      historical: {},
+    });
+    tree = renderer.create(
+      <Provider store={store}>
+        <MonthOpenCasesChart />
+      </Provider>,
+    );
+  });
+  it('should render with given state from redux store', () => {
+    expect(tree.toJSON()).toMatchSnapshot();
```
Comment (D4nielJ) [inline src/__test__/ChartOPenCases.test.js:25]: - [ ] Good job implementing new tests, however, as the previous code reviewer mentioned. It's a good idea to not only implement snapshots, as this only tests for integration reliability but not functionality. Let's also implement some unit tests for functionality.

> App has **unit tests** for several components of the app and some **pure functions** in Redux code.

File: src/components/Recovered/ChartRecoveredMonth.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  let daysOfMonth = [];
+  const newCasesData = [];
+  if (Object.keys(dataOfMonthState).length > 0) {
+    const dataOfMonth = dataOfMonthState.dates;
+    daysOfMonth = Array.from(Object.keys(dataOfMonth));
+    daysOfMonth.forEach((dayDate) => {
+      const dayData = dataOfMonth[dayDate].countries.Tunisia;
+      newCasesData.push(parseInt(dayData.today_new_recovered, 10));
+    });
+  }
```
Comment (D4nielJ) [inline src/components/Recovered/ChartRecoveredMonth.js]: For instance, you might want to test if you are getting the correct information of `newCasesData` every time.

File: src/components/Recovered/Recovered.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    totalRecovered = updateData.today_recovered;
+    if (Object.keys(dayBeforeData).length > 0) {
+      dayBeforeData = dayBeforeData.dates[previousDay].countries.Tunisia;
+      const newRecoveredDayBefore = dayBeforeData.today_new_recovered;
+      growingRate = (
+        (newRecovered - newRecoveredDayBefore) / Math.max(newRecoveredDayBefore, 1)
+      ) * 100;
+      growingRate = Math.floor(growingRate);
+    }
+  }
```
Comment (D4nielJ) [inline src/components/Recovered/Recovered.js:32]: Or you might want to make sure that the calculation of the growingRate it's correct. I think there are plenty of tests that you can implement. So please implement some, it's a good practice and very valuable in almost every company.

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { Provider } from 'react-redux';
+import store from './redux/configureStore';
+import App from './App';
+
+ReactDOM.render(
+  <React.StrictMode>
+    <Provider store={store}>
+      <App />
```
Comment (D4nielJ) [inline src/index.js:10]: [OPTIONAL] Also, I'm sorry my partner did not mention it, but there is also a requirement to implement a stateless filter component that allows the user to filter data from the app somehow.

> Uses a Filter stateless component to filter data retrieved from the API.

I highly encourage you to do this, as it's a good exercise and part of the requirements.

**Review by D4nielJ — CHANGES_REQUESTED** [review]
## Hello @Taher-web-dev, 

Great work on making the changes requested by the previous reviewer(s) 👏

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## Happy coding and stay awesome! 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can always find me on slack!

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by rukundoeric — APPROVED** [review]
Hi @Taher-web-dev,

_Excellent job on this project. Keep it up 🎆_

_Congratulations! 🎉  on the splendid job 👍
You can now merge this PR and submit the project completion form 🚀 . I wish you all the very best in your next tasks 🥇_


Cheers and Happy coding!👏👏👏

__Feel free to leave any questions or comments in the PR thread if something is not 100% clear.__

------
>As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**